File: ios/LeanIOS/Ads/AdsManager.swift
import Foundation
import UIKit

/// Public interface used from Objective-C via the generated `GonativeIO-Swift.h` header.
@objc public protocol AdsManaging {
    @objc func start(with config: AdConfig)
    @objc func showInterstitial(from vc: UIViewController)
    @objc func showRewarded(from vc: UIViewController, placement: String?, reward: @escaping (String, Int) -> Void)
    @objc func showAppOpen(from vc: UIViewController)
}

@objc public class AdConfig: NSObject {
    @objc public let admobAppId: String
    @objc public let admobInterstitial: String
    @objc public let admobRewarded: String
    @objc public let admobAppOpen: String
    @objc public let applovinSdkKey: String
    @objc public let applovinInterstitial: String
    @objc public let applovinRewarded: String
    @objc public let applovinAppOpen: String

    @objc public init(dict: [String: Any]) {
        let ads = (dict["services"] as? [String: Any])?["ads"] as? [String: Any] ?? [:]
        let am = ads["admob"] as? [String: Any] ?? [:]
        let al = ads["applovin"] as? [String: Any] ?? [:]
        func s(_ d: [String: Any], _ k: String) -> String { (d[k] as? String) ?? "" }
        admobAppId = s(am, "appId")
        admobInterstitial = s(am, "interstitial")
        admobRewarded = s(am, "rewarded")
        admobAppOpen = s(am, "appOpen")
        applovinSdkKey = s(al, "sdkKey")
        applovinInterstitial = s(al, "interstitial")
        applovinRewarded = s(al, "rewarded")
        applovinAppOpen = s(al, "appOpen")
    }
}

@objc public class AdsHub: NSObject {
    @objc public static let shared = AdsHub()
    @objc public var manager: AdsManaging?

    @objc public func bootstrap(configDict: [String: Any]) {
        let cfg = AdConfig(dict: configDict)
        #if AD_NETWORK_ADMOB
        let m: AdsManaging = AdMobAdsManager()
        #elseif AD_NETWORK_APPLOVIN
        let m: AdsManaging = AppLovinAdsManager()
        #else
        let m: AdsManaging = NoopAdsManager()
        #endif
        m.start(with: cfg)
        self.manager = m
    }
}

final class NoopAdsManager: NSObject, AdsManaging {
    func start(with config: AdConfig) {}
    func showInterstitial(from vc: UIViewController) {}
    func showRewarded(from vc: UIViewController, placement: String?, reward: @escaping (String, Int) -> Void) {}
    func showAppOpen(from vc: UIViewController) {}
}

File: android/app/src/main/java/ads/AdsManager.kt
package co.median.android.ads

import android.app.Activity
import android.app.Application

/**
 * Ad-network-agnostic contract. A concrete implementation lives in each
 * flavor source set (src/admob/ and src/applovin/).
 */
interface AdsManager {
    fun init(app: Application)
    fun preload()
    fun showInterstitial(activity: Activity, onClosed: (() -> Unit)? = null)
    fun showRewarded(activity: Activity, placement: String?, onReward: (String, Int) -> Unit, onClosed: (() -> Unit)? = null)
    fun showAppOpen(activity: Activity)
}

object Ads {
    @Volatile private var instance: AdsManager? = null

    fun get(): AdsManager = instance
        ?: error("AdsManager not initialised. Call Ads.init() from Application.onCreate().")

    fun init(app: Application, impl: AdsManager) {
        instance = impl
        impl.init(app)
    }
}

data class AdConfig(
    val admobAppId: String = "",
    val admobInterstitial: String = "",
    val admobRewarded: String = "",
    val admobAppOpen: String = "",
    val applovinSdkKey: String = "",
    val applovinInterstitial: String = "",
    val applovinRewarded: String = "",
    val applovinAppOpen: String = "",
) {
    companion object {
        fun fromAppConfigJson(json: org.json.JSONObject?): AdConfig {
            val ads = json?.optJSONObject("services")?.optJSONObject("ads") ?: return AdConfig()
            val am = ads.optJSONObject("admob") ?: org.json.JSONObject()
            val al = ads.optJSONObject("applovin") ?: org.json.JSONObject()
            return AdConfig(
                admobAppId = am.optString("appId"),
                admobInterstitial = am.optString("interstitial"),
                admobRewarded = am.optString("rewarded"),
                admobAppOpen = am.optString("appOpen"),
                applovinSdkKey = al.optString("sdkKey"),
                applovinInterstitial = al.optString("interstitial"),
                applovinRewarded = al.optString("rewarded"),
                applovinAppOpen = al.optString("appOpen"),
            )
        }
    }
}

internal object AdsConfigHolder {
    @Volatile var config: AdConfig = AdConfig()
}